lib/chef/resource/chocolatey_installer.rb

```diff
@@ -0,0 +1,198 @@
+          chocolatey_installer 'latest' do
+            action :upgrade
+            chocolatey_version '2.12.24'
+          end
+          ```
+      DOC
+
+      allowed_actions :install, :uninstall, :upgrade
+
+      property :download_url, String,
+        description: "The URL to download Chocolatey from. This defaults to the value of $env:ChocolateyDownloadUrl, if it is set, and otherwise falls back to the official Chocolatey community repository to download the Chocolatey package. It can be used for offline installation by providing a path to a Chocolatey.nupkg."
+
+      property :chocolatey_version, String,
+        description: "Specifies a target version of Chocolatey to install. By default, the latest stable version is installed. This will use the value in $env:ChocolateyVersion by default, if that environment variable is present. This parameter is ignored if download_url is set."
+
+      property :use_native_unzip, [TrueClass, FalseClass], default: false,
+        description: "If set, uses built-in Windows decompression tools instead of 7zip when unpacking the downloaded nupkg. This will be set by default if use_native_unzip is set to a value other than 'false' or '0'. This parameter will be ignored in PS 5+ in favour of using the Expand-Archive built in PowerShell cmdlet directly."
+
+      property :ignore_proxy, [TrueClass, FalseClass], default: false,
+        description: "If set, ignores any configured proxy. This will override any proxy environment variables or parameters. This will be set by default if ignore_proxy is set to a value other than 'false' or '0'."
+
+      property :proxy_url, String,
+        description: "Specifies the proxy URL to use during the download."
+
+      property :proxy_user, String,
+        description: "The username to use to build a proxy credential with. Will be consumed by the proxy_credential property if both this property and proxy_password are set"
+
+      property :proxy_password, String,
+        description: "The password to use to build a proxy credential with. Will be consumed by the proxy_credential property if both this property and proxy_user are set"
+
+      load_current_value do
+        current_state = is_choco_installed?
+        current_value_does_not_exist! if current_state == false
+        current_state
+      end
+
+      def is_choco_installed?
+        ::File.exist?("#{ENV["ALLUSERSPROFILE"]}\\chocolatey\\bin\\choco.exe")
+      end
+
+      def get_choco_version
+        powershell_exec("choco --version").result
+      end
+
+      def existing_version
+        Gem::Version.new(get_choco_version)
+      end
+
+      def define_resource_requirements
+        [ new_resource.proxy_user, new_resource.proxy_password ].each do
+          requirements.assert(:install, :upgrade).each do |a|
+            a.assertion do
+              (!new_resource.proxy_user.nil? && new_resource.proxy_password.nil?) || (new_resource.proxy_user.nil? && !new_resource.proxy_password.nil?)
```

Comment:
proxy_user.nil?|proxy_password.nil?|evaluate
       ---|       ---|---
       nil|       nil|false
       nil|   not nil|true
   not nil|       nil|true
   not nil|   not nil|false

Reply:
OOh, thanks